snap/snapcraft.yaml

```diff
@@ -0,0 +1,147 @@
+name: livebook
+base: core20
+version: "0.6.3"
+summary: Interactive & collaborative code notebooks in Elixir
+description: |
+  Livebook is a web application for writing interactive and collaborative code notebooks for Elixir.
+
+grade: devel
+confinement: strict
+architectures:
+  - build-on: amd64
+  - build-on: arm64
+
+package-repositories:
+  - type: apt
+    url: http://binaries.erlang-solutions.com/debian
+    key-id: "434975BD900CCBE4F7EE1B1ED208507CA14F4FCA"
+    suites:
+      - "focal"
+    components:
+      - "contrib"
+    architectures:
+      - "amd64"
+      - "arm64"
+
+hooks:
+  configure:
+    plugs: [network, network-bind]
+
+apps:
+  livebook:
+    extensions: [gnome-3-38]
+    command: bin/livebook-indicator
+    plugs:
+      - desktop
+      - gsettings
+      - wayland
+      - x11
+
+  server:
+    command: wrappers/start-livebook.sh
+    plugs:
+      - network-bind
+      - network
+      - home
+    environment:
+      "LIVEBOOK_HOME": $SNAP_REAL_HOME
+      "LIVEBOOK_DATA_PATH": $SNAP_USER_COMMON
+
+layout:
+  /usr/lib/erlang:
+    symlink: $SNAP/usr/lib/erlang
+  /usr/bin/git:
+    symlink: $SNAP/usr/bin/git
+
+parts:
+  wrappers:
+    plugin: dump
+    source: snap/local/wrappers
+    organize:
+      "*": "wrappers/"
+
+  elixir:
+    plugin: nil
+    source: https://github.com/elixir-lang/elixir
+    source-type: git
+    source-tag: v1.14.0-rc.1
+
+    override-build: |
+      PREFIX=$SNAPCRAFT_PART_INSTALL make install
+
+    build-packages:
+      - esl-erlang
+
+    stage-packages:
+      - esl-erlang
+      - libodbc1
+      - libglu1-mesa
+      - libwxgtk-webview3.0-gtk3-0v5
+
+  livebook:
+    after:
+      - elixir
+    source: https://github.com/livebook-dev/livebook
+    source-type: git
+    source-tag: v0.6.3
+    plugin: nil
+    override-build: |
+      mix local.hex --force && mix local.rebar --force && \
+      mix do deps.get, deps.compile && \
+      MIX_ENV=prod mix do compile, release livebook && \
+      mkdir -p $SNAPCRAFT_PART_INSTALL && \
+      cp -r _build/prod/rel/livebook $SNAPCRAFT_PART_INSTALL/livebook && \
+      echo $(cat /dev/urandom | env LC_ALL=C tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1) > $SNAPCRAFT_PART_INSTALL/livebook/releases/COOKIE
+    stage:
+      - livebook
+    build-packages:
+      - build-essential
+      - gnupg2
+    stage-packages:
+      - libcrypt1
+      - libtinfo6
+      # The following staged dependencies are inspired by the livebook Dockerfile
+      # ... as are the comments.
+      # Runtime dependencies
+      - build-essential
+      - ca-certificates
+      - libncurses5-dev
+      - git # In case someone uses `Mix.install/2` and point to a git repo
+      - wget # Additional standard tools
+      - cmake # In case someone uses Torchx for Nx
```

Comment:
> speaking of, we were thinking of including nmake.exe in Livebook for Windows exactly to be able to build certain packages but have concluded that nmake is not enough, it is really compiler toolchain and 3rd party libraries that are the sources of most friction so by shipping nmake we'd be fixing just a small part of the problem.

> double checking, is cmake (and build-essential which we have above) enough to build torchx? Don't we need any extra CUDA stuff?

Hmm, needs testing, I must confess I did not test this and really just based those packages on the Dockerfile and a cursory attempt at loading something that was fetched using git. Is there perhaps a notebook around somewhere, or an Elixir fragment, that would include all kinds of imaginable tricky package fetching scenarios that should be handled?

> Does it mean that unless we add some libX ourselves to the Snap, users won't be able to use it in their notebooks?

In short, not quite true.

Whilst I packaged this package "strictly confined", i.e. the applications inside it are executed in a containerised environment ("sandboxed") with their access to the host system limited to whatever interfaces are provided to them, and indeed filesystem access (including for system libraries) limited. The package is built with its own bunch of libraries and tools, this is true and it contributes to it being portable across Linux distributions and secure, but it also has locations under `$SNAP_DATA` and `$SNAP_USER_DATA` which are writable to it (and I certainly made an attempt to map for example `$LIVEBOOK_DATA_PATH` to one of these writable locations, hopefully an appropriate one for the purpose) and the `network` interface is also made available to the livebook server, so fetching data over the network is available to it. An analog I'd give is a Docker container with readwrite mounted volumes available to it (an application running in the container could indeed dynamically link libraries from volumes available to it).

So, the contained application could write for example dynamic libraries or executables to these locations it has access to (as a result of fetching some packages with mix for example) and then execute them, so it's in this sense not similar to the iOS and macOS sandbox where in addition to the host filesystem there's additional limitations to dynamically linking libraries at runtime and executable memory pages (~ JIT compilers).

(As far as snap startup speed, there's a lot of ongoing work on this, with this piece a good summary from earlier https://www.omgubuntu.co.uk/2022/05/ubuntu-making-firefox-snap-app-faster, and there's since been more improvements delivered still, some CPU microarchitecture agnostic, some for example ARM specific).